### PR TITLE
Add doc for scratchpad tools

### PR DESCRIPTION
**Pull request by mingshl**
### Description
Add doc for scratchpad tools

### Issues Resolved
https://github.com/opensearch-project/documentation-website/issues/11146 

### Version
_List the OpenSearch version to which this PR applies, e.g. 3.3

### Frontend features
_If you're submitting documentation for an OpenSearch Dashboards feature, add a video that shows how a user will interact with the UI step by step. A voiceover is optional._ 

### Checklist
- [ ] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by kolchfa-aws** [comment]
@mingshl  Thank you for the PR! When you connect it to an issue, you have to leave "Closes", "Fixes", or "Resolves" so it's properly linked (in this case, Closes #11146). I updated the PR description.

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+`return_history` | Boolean | Optional | When set to `true`, returns the full scratchpad content after writing. When `false` or omitted (default), returns the newly added note with confirmation
+
+## Execute parameters
+
+When executing an agent with scratchpad tools, provide the following parameters:
+
+Parameter | Type | Required/Optional | Description
+:--- | :--- | :--- | :---
+`question` | String | Required | The research question or task for the agent to work on
+`notes` | String | Optional | Specific notes to write to the scratchpad (when using WriteToScratchPadTool directly)
```
[the remark this answers is not in the document]
Comment (mingshl) [reply _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]: added example

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        }
+    ]
+}
+```
+
+### WriteToScratchPadTool
+
+Parameter | Type | Required/Optional | Description
+:--- | :--- | :--- | :---
+`notes` | String | Required (at execution) | The content to write to the scratchpad
```
Comment (nathaliellenaa) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]: why `notes` parameter here is marked as required, but on line 107, `notes` is optional?

Comment (mingshl) [reply _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]: the notes field is optional in registering the tool but it's required when executing

Comment (nathaliellenaa) [reply _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]: +1 Or we can combine into one table and mentioned optional during registration, required during execution

Comment (mingshl) [reply _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]: @pyek-bot remembered that the tool has strict schema, so the notes field is required when executing, it can be empty string but it's a required field by schema. I have to make two tables to explain when register in agents, the `notes` field is optional, but in executing the tools by api, it's required. 

check out this new commit and see this is easier to understand.

Comment (mingshl) [reply _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]: @nathaliellenaa I am sure if this is a bug in execute API.

when I check the trace of agent execution, the tool can have no parameters, but if you use the same tool with tool execute api, you cannot pass empty parameters, I think there is strong validations there. If there is, I think we should take this validation off, there are use cases of tool execution will no parameter. 

Screenshot 2025-10-10 at 4 59 53 PM

Screenshot 2025-10-10 at 4 58 47 PM

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+You can see the full content in the response:
+
+```json
+{
+    "inference_results": [
+        {
+            "output": [
+                {
+                    "name": "response",
+                    "result": "Scratchpad updated. Full content:\n- Research Plan for OpenSearch History and ML Evolution:\\n\\n1. OpenSearch version history, major releases after v2.0\\n2. For each major release:\\n    a. Key architectural upgrades\\n    b. New machine learning capabilities, especially ML Commons Agent framework \\n    c. Descriptions of major Agent tools added\\n    d. GitHub issue IDs tied to Agent framework features\\n3. Look for official OpenSearch documentation, release notes, blogs\\n4. Search code repositories for more technical details on ML changes\\n"
```
[the remark this answers is not in the document]
Comment (mingshl) [reply _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]: if user run tool execute api with return_history directly, they will never get history, so the placeholder won't be filled up. I rather showing some real response that the user can follow and get the same outcome when testing using tool execute API. 

In the later example using the scratchpad tool, then they will have some history to add to the scratchpad.

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md (hunk 2 above)
Comment (nathaliellenaa) [reply _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]: Thanks Mingshi for catching this, will take a look

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+<!-- vale off -->
+# Scratchpad tools
+**Introduced 3.3**
+{: .label .label-purple }
+<!-- vale on -->
+
+The scratchpad tools consist of `WriteToScratchPadTool` and `ReadFromScratchPadTool`, which enable agents to store and retrieve intermediate thoughts and results during runtime. These tools serve as temporary memory for a single agent execution session, allowing agents to take notes and store important findings during tool executions.
+
+**Important**: The scratchpad acts as runtime memory that persists only during a single agent execution. When you call the agent's `_execute` API, a new scratchpad is created for that session. All notes and data, except persistent_notes, are cleared when the execution completes, ensuring each execution starts with a fresh scratchpad.
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
**Important**: The scratchpad acts as runtime memory that persists only during a single agent execution. When you call the agent's `_execute` API, a new scratchpad is created for that session. All notes and data, except `persistent_notes`, are cleared when the execution completes, ensuring each execution starts with a fresh scratchpad.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+
+## Use cases
+
+- **Task decomposition**: Store research plans, intermediate findings, and progress notes during multi-step operations within a single execution
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
- **Task decomposition**: Store research plans, intermediate findings, and progress notes during multi-step operations within a single execution.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+- **Temporary state management**: Maintain context and accumulated knowledge during the current agent execution session
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
- **Temporary state management**: Maintain context and accumulated knowledge during the current agent execution session.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+- **Multi-step workflows**: Save key findings after searches to build comprehensive responses in complex tasks
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
- **Multi-step workflows**: Save key findings after searches to build comprehensive responses in complex tasks.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+- **Execution planning**: Store and reference step-by-step plans during complex operations
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
- **Execution planning**: Store and reference step-by-step plans during complex operations.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+
+## Tool parameters
+
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
The following are the parameters for the scratchpad tools.

```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+### ReadFromScratchPadTool
+
+**Registration parameters** (when adding to an agent):
+
+Parameter | Type | Required/Optional | Description
+:--- | :--- | :--- | :---
+`persistent_notes` | String | Optional | Initial notes or instructions to store in the scratchpad when first created
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
`persistent_notes` | String | Optional | Initial notes or instructions to store in the scratchpad when first created.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+
+**Execution parameters** (when calling the tool directly):
+
+Parameter | Type | Required/Optional | Description
+:--- | :--- |:------------------| :---
+`persistent_notes` | String | Required          | Initial notes or instructions to store in the scratchpad
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
`persistent_notes` | String | Required          | Initial notes or instructions to store in the scratchpad.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+- **Task decomposition**: Store research plans, intermediate findings, and progress notes during multi-step operations within a single execution
+- **Temporary state management**: Maintain context and accumulated knowledge during the current agent execution session
+- **Multi-step workflows**: Save key findings after searches to build comprehensive responses in complex tasks
+- **Execution planning**: Store and reference step-by-step plans during complex operations
+
+## Tool parameters
+
+### ReadFromScratchPadTool
+
+**Registration parameters** (when adding to an agent):
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
The following are the **registration parameters** used when adding to an agent.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+
+Parameter | Type | Required/Optional | Description
+:--- | :--- | :--- | :---
+`persistent_notes` | String | Optional | Initial notes or instructions to store in the scratchpad when first created
+
+**Execution parameters** (when calling the tool directly):
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
The following are the **execution parameters** used when calling the tool directly.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+
+Parameter | Type | Required/Optional | Description
+:--- | :--- |:------------------| :---
+`persistent_notes` | String | Required          | Initial notes or instructions to store in the scratchpad
+
+
+## Testing the tools
+
+You can directly use the tools API to execute both scratchpad tools and test their responses before registering them with your agents.
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
You can use the Tools API directly to execute both scratchpad tools and test their responses before registering them with your agents.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+
+### Testing ReadFromScratchPadTool
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
### Testing the ReadFromScratchPadTool
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+POST /_plugins/_ml/tools/_execute/ReadFromScratchPadTool
+{
+  "parameters": {
+    "persistent_notes": "You are a helpful researcher to conduct searches in OpenSearch cluster. Before making the search, please remember to use the listIndexTool to figure out what are the available indices first. When using listIndexTool, remember the index name has to be in an array format. Please write down important notes after tool used."
+  }
+}
+```
+{% include copy-curl.html %}
+
+When giving persistent_notes, you will see when tool response, it will try to show the persistent notes given in the response.
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
When provided `persistent_notes`, the tool will attempt to show the persistent notes in the response:
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+                    "name": "response",
+                    "result": "Notes from scratchpad:\n- You are a helpful researcher to conduct searches in OpenSearch cluster. Before making the search, please remember to use the listIndexTool to figure out what are the available indices first. When using listIndexTool, remember the index name has to be in an array format. Please write down important notes after tool used."
+                }
+            ]
+        }
+    ]
+}
+```
+
+You can also test with empty persistent notes.
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
You can also test with an empty `persistent_notes` field:
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+```json 
+POST /_plugins/_ml/tools/_execute/ReadFromScratchPadTool
+{
+  "parameters": {
+    "persistent_notes": ""
+  }
+}
+```
+
+You will get a response about empty scratchpad.
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
The response will indicate that the scratchpad is empty:
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+                }
+            ]
+        }
+    ]
+}
+```
+
+### WriteToScratchPadTool
+
+**Registration parameters** (when adding to an agent):
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
The following **registration parameters** are used when adding to an agent.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+
+Parameter | Type | Required/Optional | Description
+:--- | :--- | :--- | :---
+`return_history` | Boolean | Optional | When set to `true`, returns the full scratchpad content after writing. When `false` or omitted (default), returns the newly added note with confirmation
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
`return_history` | Boolean | Optional | When set to `true`, returns the full scratchpad content after writing. When `false` or omitted (default), returns the newly added note with confirmation.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 21]
```diff
@@ -0,0 +1,329 @@
...
+
+**Execution parameters** (when calling the tool directly):
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
The following **execution parameters**  are used when calling the tool directly.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 22]
```diff
@@ -0,0 +1,329 @@
...
+
+Parameter | Type | Required/Optional | Description
+:--- | :--- | :--- | :---
+`notes` | String | Required | The content to write to the scratchpad
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
`notes` | String | Required | The content to write to the scratchpad.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 23]
```diff
@@ -0,0 +1,329 @@
...
+`return_history` | Boolean | Optional | When set to `true`, returns the full scratchpad content after writing. When `false` or omitted (default), returns the newly added note with confirmation
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
`return_history` | Boolean | Optional | When set to `true`, returns the full scratchpad content after writing. When `false` or omitted (default), returns the newly added note with confirmation.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 24]
```diff
@@ -0,0 +1,329 @@
...
+
+### Testing WriteToScratchPadTool
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
### Testing the WriteToScratchPadTool
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 25]
```diff
@@ -0,0 +1,329 @@
...
+
+You can directly use the tools API to execute WriteToScratchPadTool and test the tool response before registering it with your agents.
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
You can use the Tools API directly to execute the `WriteToScratchPadTool` and test the tool response before registering it with your agents.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 26]
```diff
@@ -0,0 +1,329 @@
...
+POST /_plugins/_ml/tools/_execute/WriteToScratchPadTool
+{
+  "parameters": {
+    "notes": "Research Plan for OpenSearch History and ML Evolution:\\n\\n1. OpenSearch version history, major releases after v2.0\\n2. For each major release:\\n    a. Key architectural upgrades\\n    b. New machine learning capabilities, especially ML Commons Agent framework \\n    c. Descriptions of major Agent tools added\\n    d. GitHub issue IDs tied to Agent framework features\\n3. Look for official OpenSearch documentation, release notes, blogs\\n4. Search code repositories for more technical details on ML changes\\n"
+  }
+}
+```
+{% include copy-curl.html %}
+
+You can see the sample response from the tool output as follows:
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
The following is the example response from the tool output:
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 27]
```diff
@@ -0,0 +1,329 @@
...
+        {
+          "name": "response",
+          "result": "Wrote to scratchpad: Research Plan for OpenSearch History and ML Evolution:\\n\\n1. OpenSearch version history, major releases after v2.0\\n2. For each major release:\\n    a. Key architectural upgrades\\n    b. New machine learning capabilities, especially ML Commons Agent framework \\n    c. Descriptions of major Agent tools added\\n    d. GitHub issue IDs tied to Agent framework features\\n3. Look for official OpenSearch documentation, release notes, blogs\\n4. Search code repositories for more technical details on ML changes\\n"
+        }
+      ]
+    }
+  ]
+}
+```
+you can opt to set `return_history` parameter to `true` to get the full scratchpad content after writing.
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
You can opt to set the `return_history` parameter to `true` to get the full scratchpad content after writing:
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 28]
```diff
@@ -0,0 +1,329 @@
...
+                }
+            ]
+        }
+    ]
+}
+```
+{% include copy-curl.html %}
+
+## Example: Building a research agent with scratchpad tools
+
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md:49]:
```suggestion
Use the following steps to build a research agent with scratchpad tools.

```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 29]
```diff
@@ -0,0 +1,329 @@
...
+      }
+    ]
+  }
+}
+```
+{% include copy-curl.html %}
+
+## Step 2: Register an agent with scratchpad tools
+
+Register a conversational agent that includes both scratchpad tools along with other research tools:
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
Register a conversational agent that includes both scratchpad tools and other research tools:
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 30]
```diff
@@ -0,0 +1,329 @@
...
+    "parameters": {
+        "question": "How many residents are in New York?"
+    }
+}
+```
+{% include copy-curl.html %}
+
+
+The agent will:
+1. Read from its scratchpad to check for existing relevant information (starts empty for new executions)
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
1. Read from its scratchpad to check for existing relevant information (starts empty for new executions).
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 31]
```diff
@@ -0,0 +1,329 @@
...
+2. Create and save a research plan to the scratchpad
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
2. Create and save a research plan to the scratchpad.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 32]
```diff
@@ -0,0 +1,329 @@
...
+3. Execute searches and update the scratchpad with findings
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
3. Execute searches and update the scratchpad with findings.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 33]
```diff
@@ -0,0 +1,329 @@
...
+4. Provide a comprehensive answer based on accumulated research
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
4. Provide a comprehensive answer based on accumulated research.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 34]
```diff
@@ -0,0 +1,329 @@
...
+## Viewing scratchpad activity
+
+You can monitor how the agent uses the scratchpad by examining the execution traces:
+
+```json
+GET /_plugins/_ml/memory/message/<parent_interaction_id>/traces?next_token=0
+```
+{% include copy-curl.html %}
+
+The traces will show the sequence of scratchpad reads and writes, demonstrating how the agent builds up its knowledge during the execution session.
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
The traces will show the sequence of scratchpad reads and writes, demonstrating how the agent accumulates knowledge during the execution session.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 35]
```diff
@@ -0,0 +1,329 @@
...
+
+## Scratchpad lifecycle
+
+The scratchpad follows a simple lifecycle:
+
+1. **Creation**: A new, empty scratchpad is created when an agent execution begins
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
1. **Creation**: A new, empty scratchpad is created when an agent execution begins.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 36]
```diff
@@ -0,0 +1,329 @@
...
+2. **Usage**: During execution, the agent can read from and write to the scratchpad multiple times
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
2. **Usage**: During execution, the agent can read from and write to the scratchpad multiple times.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 37]
```diff
@@ -0,0 +1,329 @@
...
+3. **Cleanup**: The scratchpad is automatically cleared when the execution completes
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
3. **Cleanup**: The scratchpad is automatically cleared when execution completes.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 38]
```diff
@@ -0,0 +1,329 @@
...
+
+Each call to the agent's `_execute` API creates a fresh scratchpad, ensuring executions are isolated from each other.
+
+## Best practices
+
+- **Structured notes**: Encourage agents to maintain organized, structured notes in the scratchpad
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
- **Structured notes**: Encourage agents to maintain organized, structured notes in the scratchpad.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 39]
```diff
@@ -0,0 +1,329 @@
...
+- **Regular updates**: Have agents update the scratchpad after each significant step or finding
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
- **Regular updates**: Have agents update the scratchpad after each significant step or finding.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 40]
```diff
@@ -0,0 +1,329 @@
...
+- **Session awareness**: Remember that scratchpad content is temporary and specific to the current execution
+- **Efficient usage**: Use the scratchpad for intermediate results that need to be referenced multiple times during execution
+
+## Related pages
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
## Related documentation
```

**Comment by natebower** [comment]
@kolchfa-aws Editorial changes committed.

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 41]
```diff
@@ -25,161 +26,25 @@ The scratchpad tools consist of `WriteToScratchPadTool` and `ReadFromScratchPadT
...
-                }
-            ]
-        }
-    ]
-}
-```
-{% include copy-curl.html %}
+- **Structured notes**: Encourage agents to maintain organized, structured notes in the scratchpad.
+- **Regular updates**: Have agents update the scratchpad after each significant step or finding.
+- **Session awareness**: Remember that scratchpad content is temporary and specific to the current execution
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
- **Session awareness**: Remember that scratchpad content is temporary and specific to the current execution.
```

File: _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md [hunk 42]
```diff
@@ -25,161 +26,25 @@ The scratchpad tools consist of `WriteToScratchPadTool` and `ReadFromScratchPadT
...
+- **Efficient usage**: Use the scratchpad for intermediate results that need to be referenced multiple times during execution
```
Comment (natebower) [inline _ml-commons-plugin/agents-tools/tools/scratchpad-tools.md]:
```suggestion
- **Efficient usage**: Use the scratchpad for intermediate results that need to be referenced multiple times during execution.
```
